xbmc/cores/VideoPlayer/DVDInputStreams/DVDFactoryInputStream.cpp

```diff
@@ -51,16 +52,38 @@ std::shared_ptr<CDVDInputStream> CDVDFactoryInputStream::CreateInputStream(IVide
     }
   }
 
+  std::string inputstreamClass = fileitem.GetProperty(STREAM_PROPERTY_INPUTSTREAMCLASS).asString();
+
   BinaryAddonBaseList addonInfos;
   CServiceBroker::GetBinaryAddonManager().GetAddonInfos(addonInfos, true /*enabled only*/, ADDON_INPUTSTREAM);
   for (auto addonInfo : addonInfos)
   {
     if (CInputStreamAddon::Supports(addonInfo, fileitem))
-      return std::shared_ptr<CInputStreamAddon>(new CInputStreamAddon(addonInfo, pPlayer, fileitem));
+    {
+      // Used to inform input stream about special Identifier;
+      std::string instanceID =
+          fileitem.GetProperty(STREAM_PROPERTY_INPUTSTREAMCLASS_IDENT).asString();
+
+      KODI_HANDLE parentInstance = nullptr;
+      /*!
+       * Check inputstream addon is also a PVR addon.
+       * If yes and the from addon wanted worker (STREAM_PROPERTY_INPUTSTREAMCLASS)
+       * is the same as this supported inputstream addon instance give him the address to
+       * parent PVR addon instance (if available).
+       */
+      if (addonInfo->IsType(ADDON_PVRDLL) &&
```

Comment:
@FernetMenta what do you think about this? is this okay?

Reply:
The `addonInfo->IsType(ADDON_PVRDLL)` could also be omitted, it was only used for additional security as only PVR currently allows inputstream use.

Reply:
I don't like it. Any PVR specific code should go away from player. All required functions can be done via inputstream. I also don't like this parentInstance dependency.  This creates something very hard to handle. Assume you have n instances of an addon that depend on another instances that gets unloaded for whatever reason. Those n instances would most likely crash because the handle is not valid anymore.
The mechanism of passing a dependency through Kodi is really ugly. If an addon instance needs to communicate to another one, this communication path should be implemented outside of Kodi.

Reply:
I remove the line, but without a parent instance becomes mostly impossible to handle PVR and inputstream together within one addon. 

Then ugly ways would have to be carried out on the addon. In addition, when the CreateInstance call is made, the addon cannot understand who it belongs to.

Is about the use of this https://github.com/xbmc/xbmc/blob/master/xbmc/addons/kodi-addon-dev-kit/include/kodi/AddonBase.h#L263-L266, where it need a parent for (what @peak3d added about the same for codec together with inputstream instance on addon)

Reply:
If I remember correctly the decoder and inputstream instances for @peak3d 's use case are actually the same instance. For PVR this is not desired. The thumbloader for example requires multiple instances if an inputstream addon. (currently thumb-loading for PVR is disabled due to this limitation).

Reply:
Here where inputstream.adaptive create, inside inputstream instance, the class of codec instance https://github.com/peak3d/inputstream.adaptive/blob/master/src/main.cpp#L3036-L3044
His codec constructor: https://github.com/peak3d/inputstream.adaptive/blob/master/src/main.cpp#L2956

The inpustream is there the master and the codec interact with him.

For PVR comes exact the same case, only that PVR is the master and Inputstream interact with PVR.
See here https://github.com/AlwinEsch/pvr.vdr.vnsi/blob/change-to-cpp/src/client.cpp#L204-L215
This way on PVR (like on inputstream.adaptive for codec) allows to open several stream at same time. Thats also a big difference to before where only one stream was possible (maybe possibly restrictions with some addons which cannot do this themselves).

Reply:
Another part what comes with C++ on PVR system is the possibility to do this
```
                                                   .--- Inputstream Instance Channel 2 (PIP)
                                                   |
Kodi ----.----- PVR client Instance to Server A ---*--- Inputstream Instance Channel 1
     |   |
     |   +----- PVR client Instance to Server B ------- Inputstream Instance Channel 32 (Screenshot)
     |   |
     |   +----- PVR client Instance to Server C ---.--- Inputstream Instance Recording (Kodi preview)
     |   .                                         |
     |   :                                         `--- Inputstream Instance Recording 33 (Playback) --.-- Codec Instance Video
     |                                                                                                 *-- Codec Instance Audio
      `-------- Another PVR addon instances ....
```

Reply:
With an addon system supporting dependencies between addons you could create complex dependency trees. Addon B depends on A, C and D depend on B, E depends on D and so on. Now if you unload A, the entire tree has to be unloaded beginning with the leaves. You also need error handling covering complex scenarios.
Another aspect to consider is versioning. PVR is already pain in the rear. If you combine inputstream and PVR into a single addon, changes of the inputstream API become very hard.

Reply:
It was thought of **yourself** to throw away the inputstream system within PVR, completely kick out the InputStreamPVR... files in the VideoPlayer and to go together with the new inpustream addon types.

Back then I asked about a small change to the VideoPlayer for PVR and you said should wait for inputstream (https://github.com/AlwinEsch/pvr.demo/commits/add-test-stream).

About API @peak3d Looks for his inputstream changes to allow compatibilities to another version.

All this addons use the inputstream way:
- pvr.vdr.vnsi (Demux)
- pvr.rtl.radiofm (Demux)
- pvr.hts (Demux)
- pvr.vbox (ReadLiveStream)
- pvr.vuplus (ReadLiveStream)
- pvr.mythtv (ReadLiveStream)
- pvr.wmc (ReadLiveStream)
- pvr.octonet (ReadLiveStream)
- pvr.argustv (ReadLiveStream)
- pvr.dvblink (ReadLiveStream)
- pvr.mediaportal.tvserver (ReadLiveStream)
- pvr.dvbviewer (ReadLiveStream)
- pvr.nextpvr (ReadLiveStream)

They can not be all separated to independent adons for PVR and inpustream.

How the addons are handled, of course, an improvement and handling of the different addons requires the work under one another. However, this is only possible when the chaos has been cleared, and also heard to have all binary addon types in the same style.

This case here does not refer to work between addon A and B, but only for work from A to A.
Furthermore, the instance management is built in such a way that the addon is only unloaded when all instances are closed.